docs/docker-for-devops.md

```diff
@@ -0,0 +1,92 @@
+
+## Docker image structure
+
+The apply service uses a single image for the 'clock' and 'worker' container instances as well as the Linux Appliction Service. Each container instance is launched with a different start command but based on a common image.
+
+The way that developers and the production service interact with the Docker image varies and consequently we have different build targets in our Dockerfile based on the same baseline image and application codebase. The most siginifcant difference is that the production image is not bundled with any of the build dependencies to save space and reduce the number of potential vulnerability sources. The dev version of the image is over 1GB in size, the production image is nearer 200MB.
+
+The diagram below shows the hierachy of the image layers used to build our final **dev-build** and **prod-build** target images that are used by the developers and Azure deployments respectively. The images are defined in the [Dockerfile](../Dockerfile)
+
+```
+               ---------------------- ruby:2.6.5-alpine ----------------------
+               |                                                             |
+               V                                                             |
+    --- common-build-env ---                                                 |
+    |                      |                                                 |
+    V                      V                                                 V
+dev-build             prod-minify ---(copy app and compiled assets)---> prod-build
+```
+
+## Makefile
+
+To make interacting with the Docker images simpler the developers have created a [Makefile](../Makefile). Some useful commands from this Makefile are summarised below:
+- `make build`: This will build the target docker image as specified in your Docker compose file.
+- `make setup`: This will build the target docker image if not already present and set up the database.
+
+The following commands all require you to have built and setup the container with the commands above before they will run.
+- `make serve`: This will run the Docker container in headless mode so that you can interact with the web interface on port 3000.
+- `make shell`: This will launch the Docker container and drop you into a shell environment to interface with the container like any other Linux environment. Run the `exit` command to return to your local terminal.
+- `make test` and `make ci.*`: These command will run the various test suites against the application.
```

Comment:
I'm getting errors with make test and ci*:

```
% make test                                                                                                               17:30:41
docker-compose run --rm web /bin/sh -c "bundle exec rake"
Starting apply-for-postgraduate-teacher-training_db_1 ... done
Starting apply-for-postgraduate-teacher-training_redis_1 ... done
RAILS_ENV=test rails webpacker:compile
sh: node: not found
sh: nodejs: not found
Node.js not installed. Please download and install Node.js https://nodejs.org/en/download/
Exiting!
rake aborted!
Command failed with status (1): [RAILS_ENV=test rails webpacker:compile...]
/app/lib/tasks/testing.rake:49:in `block in <top (required)>'
/gems/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/cli/exec.rb:74:in `load'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/cli/exec.rb:74:in `kernel_load'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/cli/exec.rb:28:in `run'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/cli.rb:465:in `exec'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/cli.rb:27:in `dispatch'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/cli.rb:18:in `start'
/usr/local/bundle/gems/bundler-2.0.2/exe/bundle:30:in `block in <top (required)>'
/usr/local/bundle/gems/bundler-2.0.2/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
/usr/local/bundle/gems/bundler-2.0.2/exe/bundle:22:in `<top (required)>'
/usr/local/bundle/bin/bundle:23:in `load'
/usr/local/bundle/bin/bundle:23:in `<main>'
Tasks: TOP => default => run_tests => compile_assets
(See full trace by running task with --trace)
make: *** [test] Error 1
```
```
% make ci.*                                                                                                             17:30:53
zsh: no matches found: ci.*
```

Reply:
You need to set RAILS_ENV to test in `.env` to run the tests.

Reply:
I have added a note to the documentation about the environment variable change required to run the tests @saliceti.